pkg/cleaner/cleaner.go

```diff
@@ -190,13 +191,15 @@ func deleteInterface(danmClient danmclientset.Interface, ep danmv1.DanmEp) {
   }
   netInfo, err := netcontrol.GetNetworkFromEp(danmClient, &ep)
   if err != nil {
-    log.Println("WARNING: Danmep:" + ep.ObjectMeta.Name + " in namespace:" + ep.ObjectMeta.Namespace + "could not be cleaned as its network could not be GET from K8s API server:" + err.Error())
+    log.Printf("WARNING: DanmEp '%s' in namespace '%s' could not be cleaned as its network could not be GET from K8s API server: %s", ep.ObjectMeta.Name, ep.ObjectMeta.Namespace, err)
     return
   }
   //TODO: this definitely need to be expanded into a framework, where network type specific cleanup operations can be plugged-in
```

Comment:
note the TODO
it would be better to expand the existing logic into the framework, rather than adding it next to it
DeleteDanmEp also has a nice Conditionfunc already implemented in the ipam package

Reply:
I have seen the TODO comment, inspected the possibilities came up with an idea in nokia/danm#241
but you downvoted it:
> I'm not sure why you are trying to add a Cleaner API to DANM core
> DANM itself hardly cares about other CNI's IPAMs. their cleaning is called as part of their CNI DEL
> this API is for releasing DANM allocated resources

In this current form I cannot expand the API into a pluginable framework. To make DeleteDanmEp pluginable should I add a 4th function parameter (Isn't 4 params too much?) or DANM should be encapsulated into a stateful object and moving danmClient function parameter into object field? That need major refactoring effort as I see, but I will not have time to do it.
I'd prefer waiting until weave managed IPAM comes into the picture, then we could come up with a more general pluginable solution. Or perhaps you could give me guidance about your view and how should I proceed?

Also you stated that cleaner should differenciate between DANM and non-DANM IPAM managed resources:
> it should be Cleaner's responsibility to call this API when needed, and call another API to clean non-DANM IPAM managed resources. thus the plugin framework also belongs to Cleaner, not here

Those APIs are separate in this PR at the moment

> DeleteDanmEp also has a nice Conditionfunc already implemented in the ipam package

Could you detail this? I could not find it in danm `ipam` package

Reply:
of course you can. basically even outlined how to do it. you just clearly don't want to

I didn't even imply to put the API back into danm core.
Calico is called rn when the allocation was done by Calico. good. except you failed to calculate with the fact that NetworkType itself does not answer the question who allocated the IP, so the current condition is not that good, tho ye works in majority of the cases
anyway, when Calico did the allocation you call a function to do Calico cleaning
you call a function to see if DANM cleanup needs to be invoked (using https://github.com/nokia/danm/blob/master/pkg/ipam/ipam.go#L379 passing the CIDR from the network you already push through the API and using the IP from the DanmEp you alrady pushed thorugh the API) then if it did then you call a function to invoke DANM cleaning (the current one)

I also don't get your point regarding the 4th parameter. you already pass through danmClient? and a client object is there where the current call is, and from where the frame will be invoked

Reply:
I updated my  solution to use `ipam.WasIpAllocatedByDanm` when determining static CNI allocated IPs

Reply:
I have separated IPv4 and IPv6 handling